### DISCUSSION TITLE
Import statement policy

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1661] [General]
Currently there is more than one way to write import statements in `rust-bitcoin`. Should we have a policy that restricts it to a single way or does it not matter? Would there be any improvement in clarity if we had such a policy?

We are using `rustfmt` config option `group_imports = "StdExternalCrate" # Default "Preserve"` which groups external crates together however we re-export a bunch of crates, both that we control and that we do not (e.g., `hashes`, `base64`). So one can write
```rust
use hashes::Hash;
use base64::display::Base64Display;
```
as well as:
```rust
use crate::hashes::Hash;
use crate::base64::display::Base64Display;
```

Should we use `use crate::foo` for all crates that we control (i.e., ones in rust-bitcoin org) and reserve external crates for crates that are fully external?

Feel free to say a friendly "your overthinking it too much Tobin" if that is the case, I will not be offended.

**Comment by Kixunil** [comment]
Shorter is more convenient, so I prefer it. `crate::foo` looks like a module, not a crate. Of course for things like `hashes::` to work the renaming has to be in `Cargo.toml`, not `lib.rs`

**Reply by tcharding** [reply]
Cool, so one vote for using `package` in the manifest and using `use hashes::Foo` codebase wide. If no one opposes that in the next few days I'll call it gospel.

**Comment by tcharding** [comment]
Done in #1730

**Comment by tcharding** [comment]
Non-uniformity has started to kreep in again (my fault) since we created `units` and `primitives`. Do we want to re-open this discussion?

This applies to `bitcoin` but the same argument can be applied to `primitives`:

Option 1:
```rust
use crate::Amount;
```
vs

Option 2:
```rust
use units::Amount;
```

My justification for starting to favour option 1 is that I started to think that downstream would copy what we do and we are explicitly trying to 'hide' `units` and `primitives` by mirroring the API.

Should we update policy to be:

- Import explicitly from defining crate _except_ for `units` and `primitives` types which can use re-exports? 

Example

```rust
use core::fmt;

use hashes::sha256;
use internals::write_err;

use crate::{Amount, FeeRate};
```

**Reply by apoelstra** [reply]
> * Import explicitly from defining crate _except_ for `units` and `primitives` types which can use re-exports?

Yep. I would go further and say "should" instead of "can".

Can we script enforcement of this somehow? Can we put it in the format bot? (I suspect in some cases it will trigger changes to the api.txt files, which is even more reason to enforce uniformity). I wonder.

**Reply by tcharding** [reply]
My $20 a month bot is on it, don't hold your breath.

**Comment by tcharding** [comment]
hmm, found an anomaly. We have chosen to claim that `bitcoin:foo` exists for all `units::foo` but `units::parse` is not re-exported. _but_ 

```rust
use crate::parse::{self, ParseIntError, PrefixedHexError, UnprefixedHexError};
```
 seems worse than 

```rust
use units::parse::{self, ParseIntError, PrefixedHexError, UnprefixedHexError};
```

**Reply by apoelstra** [reply]
I think we should rename it to `units_parse` when re-exporting, and have a rule in our bot that it's OK if a re-export exists which is renamed like this.

**Reply by tcharding** [reply]
Solved in #4886, but use `parse_int` instead.

**Comment by tcharding** [comment]
This discussion is done with once these PRs merge:

- #4886 
- #4863 
- #4885 
- #4883
